### PR TITLE
GH-36166: [C++][MATLAB] Add utility to convert UTF-8 strings to UTF-16 and UTF-16 strings to UTF-8

### LINKED ISSUE #36166: [C++][MATLAB] Add utility to convert UTF-8 strings to UTF-16 and UTF-16 strings to UTF-8
### Describe the enhancement requested

MATLAB strings are UTF-16 encoded on all platforms. We need to write a utility that converts UTF-8 strings to UTF-16. We will also need a utility that converts UTF-16 strings to UTF-8.

### Component(s)

C++, MATLAB

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

MATLAB uses UTF-16 encoded strings, but arrow uses UTF-8.  We need a way to convert between the two encodings. 

### What changes are included in this PR?

Added two new utility functions:

1. `std::string UTF16StringToUTF8(const std::basic_string<char16_t>& source)`
2. `std::basic_string<char16_t> UTF8StringToUTF16(const std::string& source)` 

### Are these changes tested?

Added two test cases to `utf8_util_test.cc`:

1. `UTF16StringToUTF8`
2. `UTF8StringToUTF16`

### Are there any user-facing changes?
No, these APIs are intended for developers.

### Future Directions

In a followup PR, we will update the MATLAB Interface source code to use these utilities when converting between UTF16 and UTF8 encoded strings.
* Closes: #36166

### CODE REVIEW

**Comment by sgilmore10** [comment]
The CI failures seem unrelated to these changes.

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/util/utf8.cc [hunk 1]
```diff
@@ -164,5 +176,21 @@ ARROW_EXPORT Result<std::string> WideStringToUTF8(const std::wstring& source) {
   }
 }
 
+ARROW_EXPORT Result<std::string> UTF16StringToUTF8(const std::u16string& source) {
```
Comment (pitrou) [inline cpp/src/arrow/util/utf8.cc:179]: For the record, `ARROW_EXPORT` is only useful on declarations, not definitions.

File: cpp/src/arrow/util/utf8_util_test.cc [hunk 2]
```diff
@@ -397,6 +397,49 @@ TEST(WideStringToUTF8, Basics) {
...
+
+  CheckOk("", u"");
+  CheckOk("foo", u"foo");
+  CheckOk("h\xc3\xa9h\xc3\xa9", u"h\u00e9h\u00e9");
+  CheckOk("\xf0\x9f\x98\x80", u"\U0001F600");
+  CheckOk("\xf4\x8f\xbf\xbf", u"\U0010FFFF");
+  CheckOk({0, 'x'}, {0, u'x'});
+
+  CheckInvalid("\xff");
+  CheckInvalid("h\xc3");
```
Comment (pitrou) [inline cpp/src/arrow/util/utf8_util_test.cc:418]: Would have been nice to add a lone surrogate test here as well?

File: cpp/src/arrow/util/utf8.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/util/utf8.cc:179]: Oh, sorry. I missed this.

Comment (sgilmore10) [reply cpp/src/arrow/util/utf8.cc:179]: Oh that's my bad. I copied the line from the header file but forgot to delete `ARROW_EXPORT`. I'll submit a followup pull request.

**Pull request merged**